#1504 Settings page

joshxg opened the pull request:
Fixes #1504 

#### NOTE: Server needs this https://github.com/sussol/msupply/pull/4531 for this to do anything

## Change summary

First iteration of a settings page

<img width="1369" alt="image" src="https://user-images.githubusercontent.com/35858975/68538426-fb2bea80-03d8-11ea-991a-dd254ab33282.png">
<img width="1365" alt="image" src="https://user-images.githubusercontent.com/35858975/68538427-02eb8f00-03d9-11ea-85bc-f098258ff006.png">

- Needed to add `valid-url` package in order to do the same validation of the url as we do initially. Otherwise, could end up in an invalid state where you can change the URL, and be locked out of the app. Our current validation uses a method from `sussol-utilities` which imports `sussol/authentication-utilities` which uses `valid-url`. The validation method cannot be gotten at through all these packages.
- In this page, before 'saving changes' a user must enter their password. This does a local check ONLY, rather than normal login which tries to check against the server first. It would be a bit of a hassle to try and get at the method to reuse it, and I didn't want to rewrite anything. Can change this later.
- Didn't want to completely remove the normal 'admin' features yet (Long press mSupply man to bring up realm explorer/export data) - I think might have to wait a bit. Don't particularly want to be in a situation where mobile is on a version which requires `isAdmin` being in the body of responses, but the server is on an older version which isn't sending it.
- Wanting to keep the old features, but not wanting to mutate them from all over and wanting to have the UI reflect updates, used different `isAdmin` and `currentUser` props
- Added UserReducer - I think we can make this far more extensible, rather than a class which is being initialized in the root (`UserAuthenticator`) and remove the blocking of feature adding above, especially if we add user permissions (which is probably coming, lets be honest)

## Testing

#### Setup: Will need to be running a server with the linked mSupply code above, in order for the server to send `isAdmin` header.

- [ ] Login with a user WITH `view/edit preferences` permission. There should be a `Settings` button in the bottom right of the menu page.
- [ ] Login with a user WITHOUT `view/edit preferences` permission. There should be NO `Settings` button in the bottom right of the menu page.
- [ ] Clicking the SETTINGS button navigates to a SETTINGS page

**Note: Can verify details by using the Realm Explorer and entering 'Setting' in the object string**
#### SETTINGS PAGE:
- [ ] SYNC URL
    - [ ] Displays the current sync url correctly
    - [ ] Clicking to edit opens a modal, whose title gives a warning
    - [ ] Opening the modal has the sync url entered already
    - [ ] You CANNOT enter an INVALID URL as defined by [RFC 3986](https://tools.ietf.org
/html/rfc3986) (Just try some whacky ones)
    - [ ] When entering a valid URI and clicking OK, is saved.
- [ ] SYNC PASSWORD
    - [ ] Does not show the current password ever
    - [ ] uses password hiding when typing
    - [ ] Entering a new password saves it 
- [ ] REALM EXPLORER
    - [ ] Navigates to the Realm Explorer page
- [ ] EXPORT DATA
    - [ ] Exports database (to downloads folder)

### Related areas to think about

First iteration, not amazing but can get ideas and improve from there, but as it is not really user facing, I think it will be fine to get this in and work on it from there

Chris-Petty commented:
Blew up when I pressed the settings button
![image](https://user-images.githubusercontent.com/7684221/68627564-ee2b0a80-0542-11ea-9366-cc8ca90d7030.png)

joshxg commented:
sorry - that is fixed in this PR https://github.com/openmsupply/mobile/pull/1517

joshxg commented:
45a641e62e6b9edd0f7871c1457b9860e083c3b7 just adding this did make it all good

Chris-Petty commented:
# Testing

- [x] Login with a user WITH `view/edit preferences` permission. There should be a `Settings` button in the bottom right of the menu page.
- [x] Login with a user WITHOUT `view/edit preferences` permission. There should be NO `Settings` button in the bottom right of the menu page.
- [x] Clicking the SETTINGS button navigates to a SETTINGS page

**Note: Can verify details by using the Realm Explorer and entering 'Setting' in the object string**
#### SETTINGS PAGE:
- [ ] SYNC URL
    - [x] Displays the current sync url correctly
    - [x] Clicking to edit opens a modal, whose title gives a warning
    - [x] Opening the modal has the sync url entered already
    - [x] You CANNOT enter an INVALID URL as defined by [RFC 3986](https://tools.ietf.org
/html/rfc3986) (Just try some whacky ones)
    - [ ] **New** Editing and saving the sync URL doesn't wipe the password **fail**
    - [x] When entering a valid URI and clicking OK, is saved.
- [x] SYNC PASSWORD
    - [x] Does not show the current password ever
    - [x] uses password hiding when typing
    - [x] Entering a new password saves it 
- [x] REALM EXPLORER
    - [x] Navigates to the Realm Explorer page
- [x] EXPORT DATA
    - [ ] Exports database (to downloads folder) **Only worked the first time** not really this PR's problem though

Chris-Petty commented on src/localization/authStrings.js:
```diff
@@ -13,6 +13,8 @@ export const authStrings = new LocalizedStrings({
     user_name: 'User Name',
     email: 'E-mail Address',
+    warning_sync_edit:
+      'WARNING:\nThis will effect syncing for this store.\nMake sure you know what you doing!',
```
Left alignment of `<Text/>` makes this warning look a bit funny, but fine for rough pass

Chris-Petty commented on src/mSupplyMobileApp.js:
```diff
@@ -205,10 +205,10 @@ class MSupplyMobileAppContainer extends React.Component {
     const {
       confirmFinalise,
-      currentUser,
+
```
```suggestion
```

joshxg replied:
356553729a5de896faadb4b389fd656b97cf7356

Chris-Petty commented on src/navigation/constants.js:
```diff
@@ -33,4 +33,6 @@ export const ROUTES = {
   STOCKTAKE_MANAGER: 'stocktakeManager',
   STOCKTAKE_EDITOR: 'stocktakeEditor',
   STOCKTAKE_EDITOR_WITH_REASONS: 'stocktakeEditorWithReasons',
+
```
I appreciate what the new lines in this object code is doing, but it's new to me and looks weird.

joshxg replied:
its good aye

Chris-Petty replied:
Dunno if I agree with trying to do with props destructuring, which is what I think you were doing in https://github.com/openmsupply/mobile/pull/1522#discussion_r344989621

joshxg replied:
nah that was just a mistake sorry

Chris-Petty commented on src/widgets/icons.js:
```diff
@@ -48,3 +49,13 @@ export const CloseIcon = React.memo(() => <IonIcon name="md-close" size={36} col
   <FAIcon name="external-link" size={16} color={SUSSOL_ORANGE} />
 ));
+
+export const Confirm = ({ style }) => <FAIcon name="check-circle" style={style} />;
```
Needs export?

Chris-Petty replied:
Do you have to separate them out from the `memo` call to get defaultProps and propTypes in there?

joshxg replied:
this is really dumb I don't know f0bd923a47539985935a86295b998d5abfaaab23

Chris-Petty commented on src/widgets/icons.js:
```diff
@@ -48,3 +49,13 @@ export const CloseIcon = React.memo(() => <IonIcon name="md-close" size={36} col
+Confirm.defaultProps = { style: { color: FINALISE_GREEN, fontSize: 40 } };
+Confirm.propTypes = { style: PropTypes.object };
+
+export const Lock = ({ style }) => <FAIcon name="lock" size={28} style={style} />;
```
Needs export?

Chris-Petty commented on src/navigation/actions.js:
```diff
@@ -30,6 +30,17 @@ import { ROUTES } from './constants';
+  NavigationActions.navigate({
+    routeName: ROUTES.SETTINGS,
+    params: {
+      title: 'Settings',
```
I was going to ignore, but you did hook up translation strings for the modals of this page. Do the title too?

joshxg replied:
just replying to all localization comments: 
6b517762f559b7b951e4db8260e2e69e7260b4fc
08df2eb70d9437e45a04cf69191e43e64e64263b

Chris-Petty commented on src/pages/SettingsPage.js:
```diff
@@ -0,0 +1,172 @@
+  const onSave = () => openModal(MODAL_KEYS.CONFIRM_USER_PASSWORD);
+
+  const editSyncURL = newSyncURL => {
+    if (!ValidUrl.isWebUri(newSyncURL)) ToastAndroid.show('Not a valid URL', ToastAndroid.LONG);
```
I like these toasts

Chris-Petty commented on src/pages/SettingsPage.js:
```diff
@@ -0,0 +1,172 @@
+  const [state, setState] = useState({
+    syncURL: UIDatabase.getSetting(SETTINGS_KEYS.SYNC_URL),
+    modalKey: '',
+    syncPassword: '',
```
I think because of this, when you just edit the URL and save you'll wipe the password to `''`, which is oof.

Actually you wouldn't even have to save the URL, just save without doing anything and kill the password.

joshxg replied:
e87efd0cddc7eb2b66d1d5124a38d2c6b6e0ebd4

Chris-Petty commented on src/pages/SettingsPage.js:
```diff
@@ -0,0 +1,172 @@
+    setState({ ...state, modalKey: '', syncPassword: newSyncPassword });
+
+  const save = enteredPassword => {
+    const currentUserPassowrd = currentUser?.passwordHash ?? '';
```
Passowrd??

joshxg replied:
hmm I don't know what the problem is here so I added a comment explaining what a passowrd is f128c10a218f02815b680534ca179d2b3c44bbbf

Chris-Petty commented on src/pages/SettingsPage.js:
```diff
@@ -0,0 +1,172 @@
+    const passwordMatch = hashPassword(enteredPassword) === currentUserPassowrd;
+    const toastMessage = passwordMatch
+      ? 'New details have been saved'
+      : 'Password did not match. New details have not been saved.';
```
Localise or naaahhh? The modal prompting this is!

Chris-Petty commented on src/pages/SettingsPage.js:
```diff
@@ -0,0 +1,172 @@
+    }
+  };
+
+  const { menuButton, menuButtonText: buttonText } = globalStyles;
```
idk why bother aliasing?

joshxg replied:
c042edcddd6c09cbf47b1c0623f88b4b590b6012

Chris-Petty commented on src/pages/SettingsPage.js:
```diff
@@ -0,0 +1,172 @@
+});
+
+const mapStateToProps = state => ({
+  currentUser: state?.user?.currentUser,
```
Just wondering it's always good to check nullish? If state is `null` then ahhhhh? currentUser `isRequired` in propTypes anyway.

joshxg replied:
i mean this is just stupidly excessive state and user are guaranteed 049e7eea57ed57df7b2fae6fd4f3b967a82158aa

joshxg commented:
Note on export data: Looks like when I added the permissions I regressed(?) and removed the time. So, now it just uses the same name and overrides. I kind of like this behaviour I don't get as many database exports when doing stuff, but happy to make an issue and fix it if you prefer the old behaviour, I'm not too fussed

Chris-Petty commented on src/pages/SettingsPage.js:
```diff
@@ -0,0 +1,175 @@
+
+        UIDatabase.update('Setting', {
+          key: SETTINGS_KEYS.SYNC_SITE_PASSWORD_HASH,
+          value: syncPassword ? hashPassword(syncPassword) : currentUserPassword,
```
Would it be better to wrap the whole `UIDatabase.update` call an if statement with this condition instead? I guess this is less lines of code and a redundant method call is probably negligible in this case.

Chris-Petty commented:
Wait Export data opens realm explorer and vice versa

Chris-Petty commented:
Other than that seems requests/bugs have been fixed 👍

raviSussol commented:
> Wait Export data opens realm explorer and vice versa

@joshxg Yes I found the same as well. Also realm explorer doesn't seem to search entered table record(s)..

raviSussol commented:
P.S: vscode debugger on this branch still didn't work :-D Tried to copy codes from #1534 but no luck.

joshxg commented:
9bec7a2 - reversed button strings. I didn't have any problem searching in the realm explorer

Rebecca367 commented:
## Testing

#### Setup: Will need to be running a server with the linked mSupply code above, in order for the server to send `isAdmin` header.

- [x] Login with a user WITH `view/edit preferences` permission. There should be a `Settings` button in the bottom right of the menu page.
- [Failed] Login with a user WITHOUT `view/edit preferences` permission. There should be NO `Settings` button in the bottom right of the menu page.
- [x] Clicking the SETTINGS button navigates to a SETTINGS page

**Note: Can verify details by using the Realm Explorer and entering 'Setting' in the object string**
#### SETTINGS PAGE:
- [x] SYNC URL
    - [x] Displays the current sync url correctly
    - [x] Clicking to edit opens a modal, whose title gives a warning
    - [x] Opening the modal has the sync url entered already
    - [x] You CANNOT enter an INVALID URL as defined by [RFC 3986](https://tools.ietf.org
/html/rfc3986) (Just try some whacky ones)
    - [x] When entering a valid URI and clicking OK, is saved.
- [x] SYNC PASSWORD
    - [x] Does not show the current password ever
    - [x] uses password hiding when typing
    - [x] Entering a new password saves it 
- [x] REALM EXPLORER
    - [x] Navigates to the Realm Explorer page
- [x] EXPORT DATA
    - [Failed] Exports database (to downloads folder)

There is setting icon with a user WITHOUT `view/edit preferences` permission
<img width="1440" alt="Screen Shot 2019-12-11 at 8 32 51 AM" src="https://user-images.githubusercontent.com/57238075/70563202-b02d0f00-1bf2-11ea-9da0-621cdb8d99c4.png">
<img width="1440" alt="Screen Shot 2019-12-11 at 8 34 14 AM" src="https://user-images.githubusercontent.com/57238075/70563214-b58a5980-1bf2-11ea-98bd-a1d58c4e885f.png">

 [Failed] Exports database (to downloads folder)
data been exported but no show in local downloads folder
<img width="1440" alt="Screen Shot 2019-12-11 at 8 44 14 AM" src="https://user-images.githubusercontent.com/57238075/70563293-da7ecc80-1bf2-11ea-93b0-11dbfd56581f.png">

joshxg commented:
@Rebecca367 

It will export to the downloads folder on the device.

You can try to pull the data on the device to your local file system to check

```
adb pull remote local
```

for example:

```
adb pull ./storage ~/1522-test-data
```

Rebecca367 commented:
Yes, it does. Thanks for the tip.
